### PR TITLE
'conjur' class establishes identity for Windows servers

### PR DESCRIPTION
**Pull request by micahlee**
Connected to #21

### CODE REVIEW

File: manifests/config_registry.pp [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  }
+
+  if $conjur::ssl_certificate {
+    registry_value { 'HKLM\Software\CyberArk\Conjur\SslCertificate':
+      ensure => present,
+      type   => string,
+      data   => $conjur::ssl_certificate,
+    }
+  } else {
+    $cert_file = undef
```
Comment (dividedmind) [inline manifests/config_registry.pp]: Why is this needed if we're storing the cert verbatim in the registry?

BTW, I was thinking, should we use system certificate storage instead?

File: lib/facter/conjur.rb [hunk 2]
```diff
@@ -11,8 +18,23 @@ def config
       end
     end
 
+    def config_registry
+      keyname = 'Software\\CyberArk\\Conjur'
```
Comment (dividedmind) [inline lib/facter/conjur.rb]: Can this be a constant instead? I know the environment is weird inside the `setcode` block, so I'm not actually sure :)
BTW, you don't need to double the backslashes inside single quotes afaik.

File: lib/facter/conjur.rb [hunk 3]
```diff
@@ -11,8 +18,23 @@ def config
...
+
+      values = []
+      result = Win32::Registry::HKEY_LOCAL_MACHINE.open(keyname) do |reg|
+        reg.each_value { |name, type, data| values << [name.gsub(/(.)([A-Z])/, '\1_\2').downcase, data] }
```
Comment (dividedmind) [inline lib/facter/conjur.rb]: I feel this would be clearer with a map. BTW, not sure if it's worth the trouble trying to comply with the camel case convention.

File: manifests/identity_wincred.pp [hunk 4]
```diff
@@ -2,7 +2,7 @@
   if $conjur::api_key {
 
     # Temporarily use the registry as a stand-in for WinCred
-    $client_url = $conjur::client[uri]
+    $client_url = URI($conjur::client[uri]) + 'authn'
```
Comment (dividedmind) [inline manifests/identity_wincred.pp]: Is there a good reason for this?

File: lib/conjur/config.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
+module Conjur
+  module Config
```
Comment (dividedmind) [inline lib/conjur/config.rb]: Have you checked if this all actually works? I seem to remember there was a reason why I didn't split it all in separate libs. Puppet does some bizarre stuff when loading code, trying to keep modules separated. Maybe it's not as bad anymore; if that's the case, great :)

File: manifests/init.pp [hunk 6]
```diff
@@ -1,3 +1,4 @@
+# Provides access to Conjur secrets in Puppet manifests
```
Comment (dividedmind) [inline manifests/init.pp]: That's not entirely accurate. It can do that, yes, but there's value in using the class even when you're not using `conjur::secret` (namely, establishing host identity).

File: manifests/config_registry.pp (hunk 1 above)
Comment (micahlee) [reply manifests/config_registry.pp]: > Why is this needed if we're storing the cert verbatim in the registry?

Ah, thanks. It's not needed.

Comment (micahlee) [reply manifests/config_registry.pp]: > BTW, I was thinking, should we use system certificate storage instead?

Using the Cert Manager seems like a good idea. My suggestion is that we explore that as a follow on issue in parallel with getting the CI for this one set up. It's hard to tell just how much effort is or isn't required for adding Cert Manager support. 

CC: @jvanderhoof @boazmichaely @garkler, if we have the time to do it, this seems like a worthwhile additional enhancement.

File: lib/facter/conjur.rb (hunk 3 above)
Comment (micahlee) [reply lib/facter/conjur.rb]: You mean a map of the `'CamelCase' => 'under_score'` for each value?

File: manifests/identity_wincred.pp (hunk 4 above)
Comment (micahlee) [reply manifests/identity_wincred.pp]: Do you mean the `authn` suffix?

If so, I was following what I saw here: https://github.com/cyberark/conjur-puppet/blob/4aae7d6e54310cc402ddbdd87150f509c11a7b54/lib/puppet/functions/conjur/netrc.rb#L13

AFAICT, it's not necessary.

File: lib/conjur/config.rb (hunk 5 above)
Comment (micahlee) [reply lib/conjur/config.rb]: As I mentioned in Slack, this has been working for me so far. But the fact that you've seen issues before should be a risk we test further.

Do you have any ideas on test contexts that might exhibit this? The main idea I have is to go back to earlier Puppet versions. It looks like I'm testing with `5.5.14`.

File: manifests/config_registry.pp (hunk 1 above)
Comment (jvanderhoof) [reply manifests/config_registry.pp]: @micahlee, I added this to the Detailed Feature Spec as a GH suggestion under a `Future Work` section.

File: lib/facter/conjur.rb (hunk 3 above)
Comment (dividedmind) [reply lib/facter/conjur.rb]: I actually meant something like `result = reg.map { |n, t, d| [name.gsub(...), data] }.to_h` :)

File: manifests/identity_wincred.pp (hunk 4 above)
Comment (dividedmind) [reply manifests/identity_wincred.pp]: Ah, this is because some old clients use and expect full authn url in netrc file in 'host' field. Strictly speaking this should just be a hostname (I think this is also true for wincred), but we're kind of stuck with this until we're sure there aren't any that don't match for a naked hostname (which might already be true). Since we don't have any legacy clients using wincred we might as well do it correctly from the start :)

File: lib/conjur/config.rb (hunk 5 above)
Comment (dividedmind) [reply lib/conjur/config.rb]: I honestly don't remember :) I mean, if it works, then great. But let's just keep it in mind if something breaks ;)

File: manifests/params.pp [hunk 7]
```diff
@@ -12,5 +12,5 @@
     Sensitive($token.conjur::decrypt)
   }}
   $host_factory_token = undef
-  $version = $conjur_config['version'].lest || { 4 }
+  $version = $conjur_config['version'].lest || { 5 }
```
Comment (dividedmind) [inline manifests/params.pp]: I think this requires doc update. Also, I'm not sure if we should do this lest we bump major version number. This is a potentially breaking change.

File: README.md [hunk 8]
```diff
@@ -77,14 +77,49 @@ This module provides the `conjur::secret` function, described above, and the `co
...
+
+Connection settings for Conjur are stored in the Windows Registry under the key `HKLM\Software\CyberArk\Conjur`.
+The values available to set are:
+
+| Value Name | Value Type | Description |
+|-|-|-|
+| Account | REG_SZ | Conjur account specified during Conjur setup. |
+| ApplianceUrl | REG_SZ | Conjur API endpoint. |
+| SslCertificate | REG_SZ | public Conjur SSL cert. |
+| Version | REG_DWORD | Conjure API version. Defaults to `5`. |
```
Comment (dividedmind) [inline README.md]: Typo

File: lib/facter/conjur.rb (hunk 3 above)
Comment (micahlee) [reply lib/facter/conjur.rb]: Gotcha, ok. That's what I first though, though `reg` isn't enumerable on its own, so I wasn't sure.

I'll add another method in the mix to enumerate the values so that it can use `map` here to make it more clear. Thanks!

File: manifests/identity_wincred.pp (hunk 4 above)
Comment (micahlee) [reply manifests/identity_wincred.pp]: Sounds good. I'll switch to the hostname alone.

File: lib/puppet/provider/wincred_credential/wincred.rb [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+  end
+
+  def create
+    Puppet.debug("Creating WinCred credential: #{self}")
+    write_value
+  end
+
+  def flush
+    Puppet.debug("Flushing WinCred credential: #{self}")
+    return if resource[:ensure] == :absent
```
[the remark this answers is not in the document]
Comment (micahlee) [reply lib/puppet/provider/wincred_credential/wincred.rb]: Hmm... you will certainly know better than I about Puppet expectations. I was mostly following what I observed in the `puppetlabs/registry` module [1], but I see what you're saying.

I'll do a little more reading on this.

[1] https://github.com/puppetlabs/puppetlabs-registry/blob/14370830bc9618b6f8f14f0bc56c1368c1613eef/lib/puppet/provider/registry_value/registry.rb#L66-L72

File: lib/puppet/provider/wincred_credential/wincred.rb [hunk 10]
```diff
@@ -0,0 +1,59 @@
+require 'puppet/type'
+require 'wincred/wincred' if Puppet.features.microsoft_windows?
+
+Puppet::Type.type(:wincred_credential).provide(:wincred) do
```
[the remark this answers is not in the document]
Comment (micahlee) [reply lib/puppet/provider/wincred_credential/wincred.rb]: Gotcha, with WinCred in mind, it's technically a `generic_credential`, but we could just use `credential`. Any particular lean, or other suggestions?

File: lib/wincred/wincred.rb [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+      Native.CredFree(pp_credentials) if pp_credentials
+    end
+
+    def write_credential(
+        target:,
+        username:,
+        value:
+      )
+      # Convert the unicode string to a byte array
+      value_blob = value.encode('utf-8').unpack('c*')
```
[the remark this answers is not in the document]
Comment (micahlee) [reply lib/wincred/wincred.rb]: Good point! I'll make the change.

File: lib/wincred/wincred.rb [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+      }
+    end
+
+    def read_wide_string(ptr)
+      idx = 0
+      str = ''
+      while ptr[idx, 2] != "\0\0"
+        str += ptr[idx, 2].force_encoding('UTF-16LE').encode('UTF-8')
+        idx += 2
+      end
```
[the remark this answers is not in the document]
Comment (micahlee) [reply lib/wincred/wincred.rb]: The problem I'm trying to overcome is that `ptr.to_str` doesn't work at all for the double wide strings. It only returns the first character of the string contained in the pointer.

I'm not knowledgable enough in string encoding to really know how to address your concern. I will do some additional learning, but do you also have any additional suggestions in mind?

File: lib/wincred/wincred.rb [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+    ensure
+      Native.CredFree(cred_ptr) if cred_ptr
+    end
+
+    def enumerate_credentials
+      pp_credentials = Fiddle::Pointer.new 0
+
+      # Create 4-byte integer pointer to hold the returned cred count
+      free = Fiddle::Function.new(Fiddle::RUBY_FREE, [Fiddle::TYPE_VOIDP], Fiddle::TYPE_VOID)
+      cred_count_ptr = Fiddle::Pointer.malloc(Fiddle::SIZEOF_LONG, free)
```
[the remark this answers is not in the document]
Comment (micahlee) [reply lib/wincred/wincred.rb]: Ah, thank you. I was having trouble finding consistent examples or docs on this. I appreciate you looking in the source! :)

File: lib/wincred/wincred.rb [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+      cred.AttributeCount = 0
+      cred.Attributes = nil
+      cred.TargetAlias = nil
+      cred.UserName = username.encode('utf-16le')
+
+      write_result = Native.CredWriteW(Fiddle::Pointer[cred], 0)
+
+      raise "Write to WinCred failed. Error code: #{Native.GetLastError}" if write_result.zero?
+    ensure
+      Native.CredFree(cred) if cred
```
[the remark this answers is not in the document]
Comment (micahlee) [reply lib/wincred/wincred.rb]: Oh, that's right. Thanks!

File: lib/puppet/provider/wincred_credential/wincred.rb [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+    write_value
+  end
+
+  def destroy
+    Puppet.debug("Destroying WinCred credential: #{self}")
+    WinCred.delete_credential(resource.parameter(:target).value)
+  end
+
+  def username
+    current_value[:username] || :absent
```
[the remark this answers is not in the document]
Comment (micahlee) [reply lib/puppet/provider/wincred_credential/wincred.rb]: I was following `puppetlabs/registry`'s example here [1]. I'll look up if I can find any other guidance for this, but I don't personally mind returning `nil` instead.

[1] https://github.com/puppetlabs/puppetlabs-registry/blob/14370830bc9618b6f8f14f0bc56c1368c1613eef/lib/puppet/provider/registry_value/registry.rb#L87-L89

File: manifests/params.pp (hunk 7 above)
Comment (micahlee) [reply manifests/params.pp]: I suspect we'll just take the the major version bump for this. @jvanderhoof, any thoughts?

I will update the README and changelog accordingly though. Thanks!

File: lib/facter/conjur.rb (hunk 3 above)
Comment (dividedmind) [reply lib/facter/conjur.rb]: Ah; but `each_value` is enumerable, so you can do `reg.each_value.map`.

File: manifests/identity_wincred.pp (hunk 4 above)
Comment (dividedmind) [reply manifests/identity_wincred.pp]: Thanks!

File: lib/puppet/provider/wincred_credential/wincred.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (dividedmind) [reply lib/puppet/provider/wincred_credential/wincred.rb]: Yeah, that's weird, and the comment acknowledges it. I suspect it's a historical bug that has never been revisited. I think we should keep it in line with the usual meaning of `ensure => absent`.

BTW, I think we should consider spinning off this provider into a separate module.

File: lib/puppet/provider/wincred_credential/wincred.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (dividedmind) [reply lib/puppet/provider/wincred_credential/wincred.rb]: Naming is notoriously hard :) I think I like `credential` the most. Maybe we could also implement a netrc backend in the future (especially if we do spin off this module).

File: lib/wincred/wincred.rb (hunk 12 above)
[the remark this answers is not in the document]
Comment (dividedmind) [reply lib/wincred/wincred.rb]: Well, I admit it's mostly of an theoretical importance really, since it's unlikely a legit credential would have characters out of the Basic Multilingual Plane, but still it'd be good to acknowledge the deficiency in a comment at least :)

But anyway, here's what I came up with:
```ruby
def wcsnlen ptr, maxlen = 1024
  (0...maxlen).find{|i| (ptr[i*2] | ptr[i*2 + 1]) == 0 }
end

def pwchar2str ptr
  ptr.to_str(wcsnlen(ptr) * 2).force_encoding('utf-16le')
end

File: lib/puppet/provider/wincred_credential/wincred.rb (hunk 15 above)
[the remark this answers is not in the document]
Comment (dividedmind) [reply lib/puppet/provider/wincred_credential/wincred.rb]: Ah, you're right, this seems to be the convention (for whatever reason), I've checked against user provider [1].

[1] https://github.com/puppetlabs/puppet/blob/3f51eef608d713045cd4e0a59ba9b3023938fdeb/lib/puppet/provider/user/useradd.rb.

File: lib/wincred/wincred.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (dividedmind) [reply lib/wincred/wincred.rb]: Actually, I think I was wrong. It frees `ptr`, not `data->ptr`, which is the actual pointer. Unless I'm missing something, this is a ridiculous design -- the free function can't be anything else.

Anyway, you can use `Fiddle::RUBY_FREE` directly as an argument to `.malloc`, that'll at least be a little cleaner.

File: lib/facter/conjur.rb (hunk 3 above)
Comment (micahlee) [reply lib/facter/conjur.rb]: You are correct. I must be misremembering what wasn't working before. In any case, that cleans up nicely! 😄

**Comment by micahlee** [comment]
Thanks again @dividedmind!

I've pushed commits to resolve the feedback items. Let me know if you see other items to change, otherwise I will clean these commits up and re-do the original commit messages on Monday.

Regarding the other general comments:

> I wonder if we ought to add some unit tests for the wincred stuff.
Absolutely. Hugh is working on Windows executors now for Jenkins. I've gone ahead and created a ticket for this here: #28.

> BTW, I think we should consider spinning off this provider into a separate module.
I agree with this completely, and I wanted to start that way, but it was pushed back to keep the scope down initially. I've also created an issue not to lose this: #29

File: lib/wincred/wincred.rb [hunk 16]
```diff
@@ -77,8 +76,6 @@ def write_credential(
       write_result = Native.CredWriteW(Fiddle::Pointer[cred], 0)
 
       raise "Write to WinCred failed. Error code: #{Native.GetLastError}" if write_result.zero?
-    ensure
-      Native.CredFree(cred) if cred
```
Comment (dividedmind) [inline lib/wincred/wincred.rb]: I think you ought to call `Fiddle.free` here instead.
As discussed, unfortunately it doesn't seem to clean it up on its own.

File: manifests/identity/wincred.pp [hunk 17]
```diff
@@ -2,10 +2,7 @@
 # Windows Credential Manager.
 class conjur::identity::wincred inherits conjur {
   if $conjur::api_key {
-
-    $client_url = URI($conjur::client[uri]) + 'authn'
-
-    credential { "${client_url}":
+    credential { "${conjur::client[uri]}":
```
Comment (dividedmind) [inline manifests/identity/wincred.pp]: Actually, what I meant is it should be a plain hostname, eg. `conjur.example`.

File: lib/wincred/conversion.rb [hunk 18]
```diff
@@ -0,0 +1,14 @@
+module WinCred
+  # Responsible for converting w_char pointer to Ruby string
+  module Conversion
+    class << self
+      def wcsnlen(ptr, maxlen = 1024)
+        (0...maxlen).find { |i| (ptr[i * 2] | ptr[i * 2 + 1]).zero? }
```
[the remark this answers is not in the document]
Comment (micahlee) [reply lib/wincred/conversion.rb]: All thanks to rubocop 😆

**Comment by micahlee** [comment]
Thanks @dividedmind! I've added two more commits for you to fix the latest comments. Let me know if you see other corrections or improvements to make. Thanks!
